crates/hotshot-stake-table/src/vec_based.rs

```diff
@@ -0,0 +1,436 @@
+    // type IntoIter = ark_std::slice::Iter<'a, &'a (K, U256)>;
+
+    fn register(
+        &mut self,
+        new_key: &Self::Key,
+        amount: Self::Amount,
+    ) -> Result<(), StakeTableError> {
+        match self.mapping.get(new_key) {
+            Some(_) => Err(StakeTableError::ExistingKey),
+            None => {
+                let pos = self.mapping.len();
+                self.head.push((new_key.clone(), amount));
+                self.head_total_stake += amount;
+                self.mapping.insert(new_key.clone(), pos);
+                Ok(())
+            }
+        }
+    }
+
+    fn deregister(&mut self, existing_key: &Self::Key) -> Result<(), StakeTableError> {
+        match self.mapping.get(existing_key) {
+            Some(pos) => {
+                self.head_total_stake -= self.head[*pos].1;
+                self.head[*pos].1 = U256::zero();
+                Ok(())
+            }
+            None => Err(StakeTableError::KeyNotFound),
+        }
+    }
+
+    fn commitment(&self, version: SnapshotVersion) -> Result<Self::Commitment, StakeTableError> {
+        match version {
+            // IMPORTANT: we don't support committing the head version b/c it's not finalized.
+            SnapshotVersion::Head => Err(StakeTableError::SnapshotUnsupported),
+            SnapshotVersion::EpochStart => Ok(self.epoch_start_comm),
+            SnapshotVersion::LastEpochStart => Ok(self.last_epoch_start_comm),
+            SnapshotVersion::BlockNum(_) => Err(StakeTableError::SnapshotUnsupported),
+        }
+    }
+
+    fn total_stake(&self, version: SnapshotVersion) -> Result<Self::Amount, StakeTableError> {
+        match version {
+            SnapshotVersion::Head => Ok(self.head_total_stake),
+            SnapshotVersion::EpochStart => Ok(self.epoch_start_total_stake),
+            SnapshotVersion::LastEpochStart => Ok(self.last_epoch_start_total_stake),
+            SnapshotVersion::BlockNum(_) => Err(StakeTableError::SnapshotUnsupported),
+        }
+    }
+
+    fn len(&self, version: SnapshotVersion) -> Result<usize, StakeTableError> {
+        match version {
+            SnapshotVersion::Head => Ok(self.head.len()),
+            SnapshotVersion::EpochStart => Ok(self.epoch_start.len()),
+            SnapshotVersion::LastEpochStart => Ok(self.last_epoch_start.len()),
+            SnapshotVersion::BlockNum(_) => Err(StakeTableError::SnapshotUnsupported),
+        }
+    }
+
+    fn contains_key(&self, key: &Self::Key) -> bool {
+        self.mapping.contains_key(key)
+    }
+
+    fn lookup(
+        &self,
+        version: SnapshotVersion,
+        key: &Self::Key,
+    ) -> Result<(Self::Amount, Self::LookupProof), StakeTableError> {
+        match self.mapping.get(key) {
+            Some(&pos) => match version {
+                SnapshotVersion::Head => {
+                    if pos >= self.head.len() {
+                        Err(StakeTableError::KeyNotFound)
```

Comment:
we can remove lots of code duplication, by defining a `Self::get_table(&self, version) -> Result<Vec<(K, U256)>>`, see a similar [`get_root` helper function](https://github.com/EspressoSystems/HotShot/blob/develop/crates/hotshot-stake-table/src/mt_based.rs#L178-L188) here.

the `pos>=..` check can be write there once.